signature/plugin.go

```diff
@@ -220,25 +221,42 @@ func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc notat
 		return nil, err
 	}
 
+	content := envContent.Payload.Content
 	var signedPayload notation.Payload
-	if err = json.Unmarshal(envContent.Payload.Content, &signedPayload); err != nil {
+	if err = json.Unmarshal(content, &signedPayload); err != nil {
 		return nil, fmt.Errorf("signed envelope payload can't be unmarshaled: %w", err)
 	}
 
-	// TODO: Verify plugin didnot add any additional top level payload attributes. https://github.com/notaryproject/notation-go/issues/80
-	if !descriptorPartialEqual(desc, signedPayload.TargetArtifact) {
+	if !isPayloadDescriptorValid(content, desc, signedPayload.TargetArtifact) {
 		return nil, errors.New("descriptor subject has changed")
 	}
 
 	return resp.SignatureEnvelope, nil
 }
 
-// descriptorPartialEqual checks if the both descriptors point to the same resource
-// and that newDesc hasn't replaced or overridden existing annotations.
-func descriptorPartialEqual(original, newDesc notation.Descriptor) bool {
-	if !original.Equal(newDesc) {
-		return false
-	}
+func isPayloadDescriptorValid(originalContent []byte, originalDesc, newDesc notation.Descriptor) bool {
+	return originalDesc.Equal(newDesc) &&
+		!isDescriptorAttributeAdded(originalContent) &&
+		descriptorAnnotationsPartialEqual(originalDesc, newDesc)
+}
+
+// isDescriptorAttributeAdded checks that the unmarshalled json content does not have any unexpected keys added
+func isDescriptorAttributeAdded(content []byte) bool {
+	var targetArtifactMap map[string]interface{}
+	_ = json.Unmarshal(content, &targetArtifactMap)
+	descriptor := targetArtifactMap["targetArtifact"].(map[string]interface{})
+
+	// Explicitly remove expected keys to check if any are left over
+	delete(descriptor, "mediaType")
+	delete(descriptor, "digest")
+	delete(descriptor, "size")
+	delete(descriptor, "annotations")
+
+	return len(targetArtifactMap) != 1 || len(descriptor) != 0
+}
+
+// descriptorAnnotationsPartialEqual checks that newDesc hasn't replaced or overridden existing annotations.
+func descriptorAnnotationsPartialEqual(original, newDesc notation.Descriptor) bool {
```

Comment:
Should we move this method to https://github.com/notaryproject/notation-go/blob/7ad4eca1a568c4dc7bc612b589b11ef0010301b4/notation.go#L19-L36

```
// Equal reports whether d and t points to the same content.
func (d Descriptor) isSubset(t Descriptor) bool {
	return d.equals*(t) .....
}
```